tests/test_facebook_discovery.py

```diff
@@ -0,0 +1,154 @@
+        • verify that primary, replication and foreign keys
+          are given the inclusion of automatic (metadata and annotated schema).
+        • verify that all other fields have inclusion of available (metadata and schema)
+        """
+        streams_to_test = self.expected_streams()
+
+        conn_id = connections.ensure_connection(self)
+
+        found_catalogs = self.run_and_verify_check_mode(conn_id)
+
+        # Verify stream names follow naming convention
+        # streams should only have lowercase alphas and underscores
+        found_catalog_names = {c['tap_stream_id'] for c in found_catalogs}
+        self.assertTrue(all([re.fullmatch(r"[a-z_]+",  name) for name in found_catalog_names]),
+                        msg="One or more streams don't follow standard naming")
+
+        for stream in streams_to_test:
+            with self.subTest(stream=stream):
+                catalog = next(iter([catalog for catalog in found_catalogs
+                                     if catalog["stream_name"] == stream]))
+                assert catalog  # based on previous tests this should always be found
+
+                schema_and_metadata = menagerie.get_annotated_schema(conn_id, catalog['stream_id'])
+                metadata = schema_and_metadata["metadata"]
+                schema = schema_and_metadata["annotated-schema"]
+
+                # verify there is only 1 top level breadcrumb
+                stream_properties = [item for item in metadata if item.get("breadcrumb") == []]
+                self.assertTrue(len(stream_properties) == 1,
+                                msg="There is NOT only one top level breadcrumb for {}".format(stream) + \
+                                "\nstream_properties | {}".format(stream_properties))
+
+                # BUG_1 | https://stitchdata.atlassian.net/browse/SRCE-4855
+                failing_with_no_replication_keys = {
+                    'ads_insights_country', 'adsets', 'adcreative', 'ads', 'ads_insights_region',
+                    'campaigns', 'ads_insights_age_and_gender', 'ads_insights_platform_and_device',
+                    'ads_insights_dma', 'ads_insights'
+                }
+                if stream not in failing_with_no_replication_keys:  # BUG_1
+                    # verify replication key(s)
+                    self.assertEqual(
+                        set(stream_properties[0].get(
+                            "metadata", {self.REPLICATION_KEYS: []}).get(self.REPLICATION_KEYS, [])),
+                        self.expected_replication_keys()[stream],
+                        msg="expected replication key {} but actual is {}".format(
+                            self.expected_replication_keys()[stream],
+                            set(stream_properties[0].get(
+                                "metadata", {self.REPLICATION_KEYS: None}).get(
+                                    self.REPLICATION_KEYS, []))))
```

Comment:
There's a chance for an error if `"metadata"` actually isn't a key and it falls back to `{self.REPLICATION_KEYS: None}`. Because the next get will end up fetching this `None` and `set(None)` throws an error

``` python
>>> set(None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'NoneType' object is not iterable
```

I think to fix this it would look like:
``` python
set(stream_properties[0].get("metadata", {}).get(self.REPLICATION_KEYS, [])))
```
or even
``` python
set(stream_properties[0].get("metadata", {self.REPLICATION_KEYS: []}).get(self.REPLICATION_KEYS)))
```

I like the first one and we should set it to a variable to reuse in the assert too.

Reply:
Done.